lib/internal/crypto/diffiehellman.js

```diff
@@ -281,9 +317,294 @@ function diffieHellman(options) {
+  }
+
+  if (key.algorithm.name !== baseKey.algorithm.name) {
+    throw lazyDOMException(
+      'The public and private keys must be of the same type',
+      'InvalidAccessError');
+  }
+
+  if (key.algorithm.namedCurve !== baseKey.algorithm.namedCurve)
+    throw lazyDOMException('Named curve mismatch', 'InvalidAccessError');
+
+  const bits = await new Promise((resolve, reject) => {
+    deriveBitsECDH(
+      kNamedCurveAliases[baseKey.algorithm.namedCurve],
+      key[kKeyObject][kHandle],
+      baseKey[kKeyObject][kHandle], (err, bits) => {
+        if (err) return reject(err);
+        resolve(bits);
+      });
+  });
+
+  // If a length is not specified, return the full derived secret
+  if (length === null)
+    return bits;
+
+  // If the length is not a multiple of 8, it will be truncated
+  // down to the nearest multiple of 8.
+  length = MathFloor(length / 8);
+  const { byteLength } = bits;
+
+  // If the length is larger than the derived secret, throw.
+  // Otherwise, we either return the secret or a truncated
+  // slice.
+  if (byteLength < length)
+    throw lazyDOMException('derived bit length is too small', 'OperationError');
+
+  return length === byteLength ? bits : bits.slice(0, length);
+}
+
+async function asyncDeriveBitsDH(algorithm, baseKey, length) {
+  const { 'public': key } = algorithm;
+  // Null has a specific meaning for DH
+  if (length !== null)
+    validateUint32(length, 'length');
+  if (!isCryptoKey(key))
+    throw new ERR_INVALID_ARG_TYPE('algorithm.public', 'CryptoKey', key);
+
+  if (key.type !== 'public') {
+    throw lazyDOMException(
+      'algorithm.public must be a public key', 'InvalidAccessError');
+  }
+  if (baseKey.type !== 'private') {
+    throw lazyDOMException(
+      'baseKey must be a private key', 'InvalidAccessError');
+  }
+
+  if (key.algorithm.name !== 'NODE-DH')
+    throw lazyDOMException('Keys must be DH keys', 'InvalidAccessError');
+
+  if (key.algorithm.name !== baseKey.algorithm.name) {
+    throw lazyDOMException(
+      'The public and private keys must be of the same type',
+      'InvalidAccessError');
+  }
+
+  const bits = await new Promise((resolve, reject) => {
+    deriveBitsDH(
+      key[kKeyObject][kHandle],
+      baseKey[kKeyObject][kHandle], (err, bits) => {
+        if (err) return reject(err);
+        resolve(bits);
+      });
+  });
+
+  // If a length is not specified, return the full derived secret
+  if (length === null)
+    return bits;
+
+  // If the length is not a multiple of 8, it will be truncated
+  // down to the nearest multiple of 8.
+  length = MathFloor(length / 8);
+  const { byteLength } = bits;
+
+  // If the length is larger than the derived secret, throw.
+  // Otherwise, we either return the secret or a truncated
+  // slice.
+  if (byteLength < length)
+    throw lazyDOMException('derived bit length is too small', 'OperationError');
+
+  return length === byteLength ? bits : bits.slice(0, length);
```

Comment:
this `.slice` needs to be a primordial, but i'm not sure on which constructor. one of the TypedArrays, perhaps?

Reply:
String.prototype.slice, ArrayBuffer.prototype.slice, and TypedArray.prototype.slice are not interchangeable so it'll take a bit of doing... will look into it.